app/models/ability.rb

```diff
@@ -35,17 +35,61 @@ def initialize(user)
     #   end
 
     user ||= User.new # guest user (not logged in)
-    # All users can only update their own user attributes
     user_id = user.id
+
+    # All users can only update their own user attributes
     can :update, User, id: user_id
 
+    can_assign_ownership(user)
+
+    can_delete_ownership(user)
+
+    can_delete_membership(team_member, user)
+
     can :read, Team, private: false
 
     if user.present?
+      can_crud_team(user)
+    end
+  end
+
+  private
+    def can_crud_team(user)
+      user_id = user.id
       can :create, Team
       can :read, Team, private: true, members: { id: user_id }
       can :update, Team, members: { id: user_id }
       can :destroy, Team, owners: { id: user_id }
     end
-  end
+
+    def can_assign_ownership(user)
+      can :assign_ownership, Team, Team do |team|
+        team.owners.include? user
+      end
+    end
+
+    def can_delete_membership(team_member, user)
+      can :delete_membership, Team, Team do |team|
+        user_id = user.id
+        exist_owners_after_delete = owners_after_delete = Ability.number_of_owners_after_delete(team, team_member) > 0
+        ((team.owners.include? user) && exist_owners_after_delete) || ((user_id == Integer(team_member)) && exist_owners_after_delete)
```

Comment:
- @tomkellog we had another fundamental talk right after you left, let's talk tomorrow or later :-)
- to summarize:
  - Let's try to reduce complexity of ability to only contain simple access abilities, like "can the current user access members of a specific team"
  - logic for allowing more complex actions (e.g. delete membership) will be added to the specific controller